policy/modules/services/spamassassin.te

```diff
@@ -549,10 +555,61 @@ userdom_use_inherited_user_terminals(spamd_update_t)
 userdom_dontaudit_search_user_home_dirs(spamd_update_t)
 userdom_dontaudit_search_user_home_content(spamd_update_t)
 
+tunable_policy(`spamassassin_network_update',`
+	corenet_all_recvfrom_netlabel(spamd_update_t)
+	corenet_tcp_sendrecv_generic_if(spamd_update_t)
+	corenet_tcp_sendrecv_generic_node(spamd_update_t)
+	corenet_tcp_bind_generic_node(spamd_update_t)
+	corenet_udp_bind_generic_node(spamd_update_t)
+	corenet_sendrecv_http_client_packets(spamd_update_t)
+	corenet_tcp_connect_http_port(spamd_update_t)
+')
+
 optional_policy(`
 	cron_system_entry(spamd_update_t, spamd_update_exec_t)
 ')
 
 optional_policy(`
-	gpg_exec(spamd_update_t)
+	gpg_spec_domtrans(spamd_update_t, spamd_update_gpg_t)
+')
+
+optional_policy(`
+	mta_read_config(spamd_update_t)
+')
+
+########################################
+#
+# GPG local policy
+#
+
+allow spamd_update_gpg_t self:process setrlimit;
+
+allow spamd_update_gpg_t spamd_update_t:fd use;
+allow spamd_update_gpg_t spamd_update_t:fifo_file rw_fifo_file_perms;
+allow spamd_update_gpg_t spamd_update_t:process sigchld;
+
+manage_dirs_pattern(spamd_update_gpg_t, spamd_tmp_t, spamd_tmp_t)
+manage_files_pattern(spamd_update_gpg_t, spamd_tmp_t, spamd_tmp_t)
+files_tmp_filetrans(spamd_update_gpg_t, spamd_tmp_t, { file dir })
+
+kernel_dontaudit_search_sysctl(spamd_update_gpg_t)
+
+files_read_usr_files(spamd_update_gpg_t)
+
+miscfiles_read_localization(spamd_update_gpg_t)
+
+userdom_use_unpriv_users_fds(spamd_update_gpg_t)
+userdom_use_inherited_user_terminals(spamd_update_gpg_t)
+
+optional_policy(`
+	gpg_entry_type(spamd_update_gpg_t)
+	gpg_exec(spamd_update_gpg_t)
+	gpg_dontaudit_getattr_gpg_runtime_dirs(spamd_update_gpg_t)
+	gpg_dontaudit_search_user_secrets_dirs(spamd_update_gpg_t)
+	gpg_agent_entry_type(spamd_update_gpg_t)
+	gpg_agent_exec(spamd_update_gpg_t)
+')
+
+optional_policy(`
+	mta_manage_config(spamd_update_gpg_t)
```

Comment:
I have other issues with this PR, but this is still the big one, as it was last time. I still think this domain should be dropped and the permissions added to `spamd_update_t`.

Reply:
Out of curiosity, why does `spamd_update_t` need to execute `gpg`? Are the spamassassin rules GPG-signed?

Reply:
If that is the case I second just merging this behavior into `spamd_update_t`, analogous to how I implemented this behavior for container image signature verification: https://github.com/SELinuxProject/refpolicy/blob/main/policy/modules/services/container.te#L780-L785

Reply:
Yes, the spamassassin rules are signed with GPG key whose fingerprints are embedded in the sa-update script.

Reply:
By having just one domain (spamd_update_t) instead of two (spamd_update_t and spamd_update_gpg_t) the level of security is decreased, because spamd_update_t is network-facing and if it is compromised it can potentially allow a remote attacker to access secret GPG files.

Reply:
So why do you want to drop the safer two domains design and employ just one domain for all operations ?

Reply:
> By having just one domain (spamd_update_t) instead of two (spamd_update_t and spamd_update_gpg_t for the child GPG verification process) the level of security is decreased, because spamd_update_t is network-facing and if it is compromised it can potentially allow a remote attacker to access secret GPG files.

Are there secret keys on each system or just the public key on each system?  I would hope that the secret keys are kept secret by spamassassin.

Reply:
Secret (or private) crypto keys are stored in the home directory under ~/.gnupg/secring.gpg if a user has previously created them (for encrypting emails or signing/encrypting other kinds of documents and files sent over the network).

The trust database is also critical to confidentiality and it is stored under ~/.gnupg/trustdb.gpg.

Reply:
Yes, of course I know that at the moment there is a safety margin.

However, if things can be changed so that they provide a much larger safety margin at no cost, then we end up with a better design which is safer.

Reply:
> However, if things can be changed so that they provide a much larger safety margin

I don't understand what you mean by safety margin.  The policy is denying access.

> at no cost, then we end up with a better design which is safer.

There is a cost. More types means more policy complexity. More policy complexity means increased difficulty to understand and maintain. You're saying this is a better design, but no one else seems to agree (if you're monitoring this thread but not commenting please thumbs up or thumbs down this post to indicate your opinion).

In functional terms, more policy also means more non-swappable kernel memory used.

Reply:
For safety margin, I mean stay away from sensitive domains.

dontaudit does not guarantee anything other than not ending up in the audit log.

neverallow is the only statement which guarantees avoiding a given permission.

Does it make sense ?

Reply:
Nobody else has commented, so it's up to you, I have just committed the amendments you requested with the warning already made...